armi/reactor/blueprints/__init__.py

```diff
@@ -102,6 +104,7 @@
 from armi.reactor.blueprints.componentBlueprint import ComponentGroups
 from armi.reactor.blueprints import isotopicOptions
 from armi.reactor.blueprints.gridBlueprint import Grids, Triplet
+from armi.reactor.converters.axialExpansionChanger import AxialExpansionChanger
```

Comment:
Hmmm

So, the primary blueprint code knows about the Axial Expansion?

But two-thirds of the projects that use ARMI, and half the ones at TerraPower, don't even have axial assemblies.

So, how does this code work for, say, a molten chloride reactor?

I ask for two reasons: 

1. ARMI has to work for all reactors, not just one.
2. Axial Expansion is very slow, and for sure we don't want to instantiate all that unless we have to.

Reply:
If you look at the code it’s only called if input  heights considered cold 

Before and after this change the calls are made by the time o=armi.init is done running.

Reply:
That's a fair point.  Okay, so we aren't breaking non-axial-type reactors.

But we are importing something that is specific to a TYPE of reactor (albeit a common one) into a part of the code that (right now) reads YAML files and builds some math.

It just _feels_ like something that should be in [assemblyBlueprint.py](https://github.com/terrapower/armi/blob/main/armi/reactor/blueprints/assemblyBlueprint.py).  But, I assume, it was just hard to work into that module?

Reply:
you need to know the meshes of each of the individual assemblies to go through the processes since they all have their indices mapped to the reference assembly (see `a.makeAxialSnapList(referenceAssembly)`). I dont see anywhere in that module where there are the full set of assemblies. The nice thing about doing it here is that the blueprints assemblies will exists for the shortest possible time ad the incorrect dimensions/masses since they are fixed right after construction (line 322)

Reply:
if you want, right after the area consistancy check i could move all the new code to 
`axialExpansionChanger`

and have it be
```
if not cs["detailedAxialExpansion"]:
    axialExpansionChanger.makeAssembliesSnapToUniformMesh(self.assemblies.values(), cs["nonUniformAssemFlags"])
if not cs["inputHeightsConsideredHot"]:
    axialExpansionChanger.expandColdDimsToHot(self.assemblies.values(),  cs["detailedAxialExpansion"])
```

Fundamentally though, blueprints should be responsible for producing assemblies with correct masses and elevations. It can certainly call someone else to do the work though

Reply:
1. I see the problem you're trying to solve, sure.
2. I love that you moved the axial expansion logic into that file/module.
3. Taking a step back, I think `blueprints/__init__.py` already has [direct references to assemblies](https://github.com/terrapower/armi/blob/56bb2591cbb1be65ebb2aab2cc3adf7f604a9731/armi/reactor/blueprints/__init__.py#L225-L228) at the top level.  So I think you're not causing a design flaw so much as stuck within the system we created, which is still very assembly-first. Bummer.
4. Let's explore your solution. The worst case is that each special type of reactor needs a bunch of these `if cs["settingName"]` statements, which some non-blueprint method. Unless that list gets long, that's probably going to look a bit ugly, but not be heart-breakingly bad.
5. Sure, after moving the logic to the axial expansion module, I'm convinced.

Sorry, I'm not trying to be a bother. I have just learned by hard experience that we improving something at the start saves a TON of time spent on technical debt later. I keep meaning to give a talk on that.

Reply:
Great. thanks for clarifying @john-science. I genuinely don't think this list of things will grow considerably. I would recommend describing molten salt reactors and TRISO-based fuel as "non-pin-based" or "non-assembly-based" as opposed to "non-axial-type reactors". 

Molten salt reactors still typically care about axial expansion of their vessel and the impact of that on reactivity (I spent a couple years on a molten salt project). Its just that the framework currently doesn't support it. Its not a  trivial feature though and its a bit different from a fuel pin expanding. Tt took a couple hundred hours for us to get cold to hot dims working for pin-based-reactors. I am sure that if we had the capability to input cold vessel dims and have the framework expand the vessel to hot, molten salt reactors would want to use the capability.

`Sure, after moving the logic to the axial expansion module, I'm convinced.`  what more do you need to approve the PR?

Reply:
>  I would recommend describing molten salt reactors and TRISO-based fuel as "non-pin-based" or "non-assembly-based" as opposed to "non-axial-type reactors".

Thanks.  And, always, I am interested in learning more about reactor design/analysis.  Sorry, my graduate degrees are in physics, but not the right kind. I'm happy to learn.

> what more do you need to approve the PR?

I added one little comment about a docstring, but I'm good otherwise.  The logic looks sound, and I like the new layout better.  

Thanks, man!